### PR TITLE
Done translating high-order components

### PR DESCRIPTION
**Pull request by aladin002dz**
booking the page "Higher-order components" translation

### CODE REVIEW

**Comment by aladin002dz** [comment]
Done :)

**Comment by iRayan7** [comment]
Hi @mohshbool, please review this page. Thanks.

**Comment by mohshbool** [comment]
> Done :)

Great work on your translation. I just had one comment because I think this is a big change.
I noticed that you translated 'high-order components' => 'المكونات ذات الترتيب الأعلى' but according to [the glossary](https://github.com/reactjs/ar.reactjs.org/wiki/Glossary#common-translations) it should be 'مكونات من المرتبة الأعلى'.

**Comment by 3imed-jaberi** [comment]
> > Done :)
> 
> Great work on your translation. I just had one comment because I think this is a big change.
> I noticed that you translated 'high-order components' => 'المكونات ذات الترتيب الأعلى' but according to [the glossary](https://github.com/reactjs/ar.reactjs.org/wiki/Glossary#common-translations) it should be 'مكونات من المرتبة الأعلى'.

We have faced the same problem in `docs/nav.yml` (  me & @Fcmam5  )  .. i think `المكونات ذات الترتيب الأعلى`  is better ..
> --> should we update the glossary in this point.

**Comment by aladin002dz** [comment]
Should we stick with the glossary or with hsoub translation?

**Comment by mohshbool** [comment]
> Should we stick with the glossary or with hsoub translation?

not sure but I personally went with the glossary, I actually only compared my translation with hsoub's after I was done translating, rather than basing my translation on it

**Comment by 3imed-jaberi** [comment]
> Should we stick with the glossary or with hsoub translation?

### I suggested that we stick to a wiki-hasoub at the beginning and when we finish everything ( first release ) we organize a good glossary and fix all small bugs.
### @iRayan7  @Fcmam5  @aladin002dz  @mohshbool @maskeddeveloper

**Comment by aladin002dz** [comment]
> > Should we stick with the glossary or with hsoub translation?
> 
> ### I suggested that we stick to a wiki-hasoub at the beginning and when we finish everything ( first release ) we organize a good glossary and fix all small bugs.
> ### @iRayan7 @Fcmam5 @aladin002dz @mohshbool @maskeddeveloper

This is the approach I am supporting.

File: content/docs/higher-order-components.md [hunk 1]
```diff
@@ -1,32 +1,32 @@
 ---
 id: higher-order-components
-title: Higher-Order Components
+title: المكونات ذات الترتيب الأعلى
 permalink: docs/higher-order-components.html
 ---
 
-A higher-order component (HOC) is an advanced technique in React for reusing component logic. HOCs are not part of the React API, per se. They are a pattern that emerges from React's compositional nature.
+إنّ المكوّنات ذات الترتيب الأعلى (Higher-Order Components واختصارًا HOC) هي تقنية متقدمة في React لإعادة استخدام منطق المكونات. وهي ليست جزءًا من واجهة برمجة تطبيقات React API، بل هي نمط ينبثق عن طبيعة React التركيبية.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]: just some حركات 

```suggestion
إنّ المُكوّنات ذات الترتيب الأعلى (Higher-Order Components واختصارًا HOC) هي تِقنية مُتَقَدِمة في React لإعادة استخدام مَنطِق المُكونات. وهي ليست جزءًا من واجهة برمجة تطبيقات React API، بَل هي نَمَط يَنبَثِق عن طَبيعة React التَركيبية.
```

File: content/docs/higher-order-components.md [hunk 2]
```diff
@@ -1,32 +1,32 @@
...
 
-Concretely, **a higher-order component is a function that takes a component and returns a new component.**
+باختصار، **يكون المكوّن ذو الترتيب الأعلى عبارة عن دالة تأخذ مكوّنًا وتُعيد مُكوّنًا جديدًا.**
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
باختصار، **المكوّن ذو الترتيب الأعلى هو عبارة عن دالّة تَأخُذ مُكوّنًا وتُعيد مُكوّنًا جديدًا.**
```

File: content/docs/higher-order-components.md [hunk 3]
```diff
@@ -1,32 +1,32 @@
...
 
 ```js
 const EnhancedComponent = higherOrderComponent(WrappedComponent);
 ```
 
-Whereas a component transforms props into UI, a higher-order component transforms a component into another component.
+وفيما يُحوّل المكوّن الخاصيّات إلى واجهة مستخدم، يُحوّل المكوّن ذو الترتيب الأعلى مكوّنًا إلى مكوّن آخر.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
بينما يُحوّل المكوّن الخاصيّات إلى واجهة مستخدم، يُحوِّل المُكوّن ذو الترتيب الأعلى مكوّنًا إلى مكوّن آخر.
```

File: content/docs/higher-order-components.md [hunk 4]
```diff
@@ -1,32 +1,32 @@
...
-In this document, we'll discuss why higher-order components are useful, and how to write your own.
+سنناقش في هذه الصفحة الفائدة من المكوّنات ذات الترتيب الأعلى وكيفية كتابتها.
 
-## Use HOCs For Cross-Cutting Concerns {#use-hocs-for-cross-cutting-concerns}
+## استخدام المكوّنات ذات الترتيب الأعلى لأجل الاهتمامات المشتركة {#use-hocs-for-cross-cutting-concerns}
 
 > **Note**
 >
-> We previously recommended mixins as a way to handle cross-cutting concerns. We've since realized that mixins create more trouble than they are worth. [Read more](/blog/2016/07/13/mixins-considered-harmful.html) about why we've moved away from mixins and how you can transition your existing components.
+>أشرنا سابقًا إلى أفضلية استخدام المخاليط (mixins) كطريقة للتعامل مع الاهتمامات المشتركة (cross-cutting concerns)، ولكننا أدركنا بعد ذلك أنّ المخاليط تُسبّب مشاكل أكثر من فائدتها. تعرّف من [هنا](/blog/2016/07/13/mixins-considered-harmful.html) عن سبب انتقالنا من المخاليط وكيفية تحويل مكوّناتك الحالية التي تستخدمها.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
>أشرنا سابقًا إلى أفضلية استخدام المخاليط (mixins) كطريقة للتعامل مع الاهتمامات المُشتركة (cross-cutting concerns)، ولكنّنا أدركنا بعد ذلك أنّ المخاليط تُسَبّب مشاكل أكثر من فائدتها. تعرّف من [هنا](/blog/2016/07/13/mixins-considered-harmful.html) عن سبب انتقالنا من المخاليط وكيفية تحويل مُكوّناتك الحالية التي تَستخدِمُها.
```

File: content/docs/higher-order-components.md [hunk 5]
```diff
@@ -121,10 +120,9 @@ const BlogPostWithSubscription = withSubscription(
   (DataSource, props) => DataSource.getBlogPost(props.id)
 );
 ```
+المُعامِل الأول هو المكوّن المُغلَّف. يسترجع المُعامِل الثاني البيانات التي تُهمّنا، مع إعطاء مصدر البيانات `DataSource` والخاصيّات الحاليّة.
 
-The first parameter is the wrapped component. The second parameter retrieves the data we're interested in, given a `DataSource` and the current props.
-
-When `CommentListWithSubscription` and `BlogPostWithSubscription` are rendered, `CommentList` and `BlogPost` will be passed a `data` prop with the most current data retrieved from `DataSource`:
+عند تصيير `CommentListWithSubscription` و `BlogPostWithSubscription`، فسيُمرِّر المكوّنان `CommentList` و `BlogPost` خاصيّة للبيانات `data` والتي تحمل أحدث البيانات المستخرجة من `DataSource`:
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
عند تصيير `CommentListWithSubscription` و `BlogPostWithSubscription`، فَسيُمَرَّر للمُكوّنان `CommentList` و `BlogPost` خاصيّة للبيانات `data` والتي تحمل أحدث البيانات المستخرجة من `DataSource`:
```

File: content/docs/higher-order-components.md [hunk 6]
```diff
@@ -163,17 +161,17 @@ function withSubscription(WrappedComponent, selectData) {
...
+يستقبل المكوّن المُغلَّف جميع الخاصيّات من الحاوية بالإضافة إلى الخاصيّة الجديدة وهي `data` والتي يستخدمها لتصيير ناتجه. لا يهتم المكوّن ذو الترتيب الأعلى بكيفية أو سبب استخدام البيانات، ولا يهتم المكوّن المُغلَّف بمصدر البيانات.
 
-Because `withSubscription` is a normal function, you can add as many or as few arguments as you like. For example, you may want to make the name of the `data` prop configurable, to further isolate the HOC from the wrapped component. Or you could accept an argument that configures `shouldComponentUpdate`, or one that configures the data source. These are all possible because the HOC has full control over how the component is defined.
+بما أنّ `withSubscription` هو دالة عادية بإمكانك إضافة وسائط لها كما تريد. فقد ترغب مثلًا بجعل اسم الخاصيّة `data` قابلًا للإعداد، وذلك لعزل المكوّن ذي الترتيب الأعلى عن المكوّن المُغلِّف له، أو تستطيع قبول وسيط يُعِد `shouldComponentUpdate` أو مصدر البيانات. كل هذه الإمكانيات متوفرة بسبب امتلاك المكوّن ذو الترتيب الأعلى السيطرة على كيفيّة تعريف المكوّنات.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
بما أنّ `withSubscription` دالّة عادية بإمكانك إضافة وسائط لها كما تريد. فقد ترغب مثلًا بِجَعل اسم الخاصيّة `data` قابلًا للإعداد، وذلك لزيادة عَزل المكوّن ذي الترتيب الأعلى عن المكوّن المُغلِّف له، أو تستطيع قبول وسيط يُعِد `shouldComponentUpdate` أو مصدر البيانات. كل هذه الإمكانيات متوفرة بسبب امتلاك المكوّن ذو الترتيب الأعلى السيطرة على كيفيّة تعريف المكوّنات.
```

File: content/docs/higher-order-components.md [hunk 7]
```diff
@@ -99,16 +99,15 @@ class BlogPost extends React.Component {
...
-- Inside the listener, call `setState` whenever the data source changes.
-- On unmount, remove the change listener.
+بإمكانك أن تتخيّل في التطبيقات الكبيرة تكرار هذا النمط من الاشتراك بمصدر البيانات `DataSource` واستدعاء `setState`. نريد وحدة مُجرَّدة تسمح لنا بتعريف هذا المنطق في مكان واحد ومشاركته عبر مكوّنات عديدة. وهنا تأتي فائدة المكوّنات ذات الترتيب الأعلى.
 
-You can imagine that in a large app, this same pattern of subscribing to `DataSource` and calling `setState` will occur over and over again. We want an abstraction that allows us to define this logic in a single place and share it across many components. This is where higher-order components excel.
-
-We can write a function that creates components, like `CommentList` and `BlogPost`, that subscribe to `DataSource`. The function will accept as one of its arguments a child component that receives the subscribed data as a prop. Let's call the function `withSubscription`:
+نستطيع كتابة دالة تُنشِئ مكوّنات، مثل `CommentList` و `BlogPost` والتي تشترك بمصدر البيانات `DataSource`. تقبل هذه الدالة كوسيط لها المكوّن الابن الذي يستقبل البيانات المُشارَكَة كخاصيّة له. فلنسمّي هذه الدالة `withSubscription`:
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
نستطيع كتابة دالّة تُنشِئ مكوّنات، مثل `CommentList` و `BlogPost` والتي تشترك بمصدر البيانات `DataSource`. تقبل هذه الدالّة كوسيط لها المكوّن الابن الذي يستقبل البيانات المُشارَكَة كخاصيّة له. فلنسمّي هذه الدالّة `withSubscription`:
```

File: content/docs/higher-order-components.md [hunk 8]
```diff
@@ -163,17 +161,17 @@ function withSubscription(WrappedComponent, selectData) {
 }
 ```
 
-Note that a HOC doesn't modify the input component, nor does it use inheritance to copy its behavior. Rather, a HOC *composes* the original component by *wrapping* it in a container component. A HOC is a pure function with zero side-effects.
+لاحظ أنّ المكوّن ذي الترتيب الأعلى لا يُعدِّل مكوّن حقل الإدخال ولا يستخدم الوراثة لنسخ سلوكه، بل يُركِّب المكوّن الأساسي عن طريق تغليفه في مكوّن حاوية. المكوّن ذو الترتيب الأعلى هو عبارة عن دالة نقيّة (pure) بدون أي تأُثيرات جانبية إطلاقًا.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
لاحظ أنّ المكوّن ذي الترتيب الأعلى لا يُعدِّل مكوّن حقل الإدخال ولا يستخدم الوراثة لنسخ سلوكه، بل يُركِّب المكوّن الأساسي عن طريق تغليفه في مكوّن حاوية. المكوّن ذو الترتيب الأعلى هو عبارة عن دالّة نقيّة (pure) بدون أي تأُثيرات جانبية إطلاقًا.
```

File: content/docs/higher-order-components.md [hunk 9]
```diff
@@ -190,11 +188,11 @@ function logProps(InputComponent) {
 const EnhancedComponent = logProps(InputComponent);
 ```
 
-There are a few problems with this. One is that the input component cannot be reused separately from the enhanced component. More crucially, if you apply another HOC to `EnhancedComponent` that *also* mutates `componentWillReceiveProps`, the first HOC's functionality will be overridden! This HOC also won't work with function components, which do not have lifecycle methods.
+هنالك بعض المشاكل عند فعل ذلك. أحدها هي عدم القدرة على استخدام مكوّن حقل الإدخال بشكل منفصل عن المكوّن `EnhancedComponent`. وإن طبقت مكوّن ذو ترتيب أعلى آخر إلى المكوّن `EnhancedComponent` والذي يُعدِّل *أيضًا* `componentWillReceiveProps`، فسيتجاوز وظيفة المكوّن ذو الترتيب الأعلى الأول! لا يعمل المكوّن ذو الترتيب الأعلى هذا أيضًا مع المكوّنات الدالية لأنّها لا تمتلك توابع دورة الحياة.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
هنالك بعض المشاكل عند فعل ذلك. أحدها هي عدم القدرة على استخدام مكوّن حقل الإدخال بشكل منفصل عن المكوّن `EnhancedComponent`. وإن طبقت مكوّن ذو ترتيب أعلى آخر إلى المكوّن `EnhancedComponent` والذي يُعدِّل *أيضًا* `componentWillReceiveProps`، فسيتجاوز وظيفة المكوّن ذو الترتيب الأعلى الأول! لا يعمل المكوّن ذو الترتيب الأعلى هذا أيضًا مع المكوّنات الدالّية لأنّها لا تمتلك توابع دورة الحياة.
```

File: content/docs/higher-order-components.md [hunk 10]
```diff
@@ -190,11 +188,11 @@ function logProps(InputComponent) {
...
 
-Mutating HOCs are a leaky abstraction—the consumer must know how they are implemented in order to avoid conflicts with other HOCs.
+إنّ تعديل المكوّنات ذات الترتيب الأعلى ليس أمرًا بسيطًا فيجب معرفة كيفية تنفيذها من أجل تجنب التعارض مع المكوّنات ذات الترتيب الأعلى الأخرى.
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]:
```suggestion
إنّ تعديل المكوّنات ذات الترتيب الأعلى ليس أمرًا بسيطًا فيجب معرفة كيفية تنفيذها لتَجنُب التعارض مع المكوّنات ذات الترتيب الأعلى الأخرى.
```

File: content/docs/higher-order-components.md [hunk 11]
```diff
@@ -190,11 +188,11 @@ function logProps(InputComponent) {
...
 
-Instead of mutation, HOCs should use composition, by wrapping the input component in a container component:
+بدلًا من تعديل المكوّن ذو الترتيب الأعلى يجب استخدام التراكيب عن طريق تغليف مكوّن حقل الإدخال في مكوّن حاوية:
```
Comment (mohshbool) [inline content/docs/higher-order-components.md]: not very sure about this

```suggestion
بدلًا من التعديل ، المكوّنات ذو الترتيب الأعلى يجب ان تَستخدِم التراكيب عن طريق تغليف مكوّن حقل الإدخال في مكوّن حاوية:
```

**Comment by aladin002dz** [comment]
> please allow me to make those little suggestions 😃
> i will continue the review process another time

@mohshbool all your suggestions committed! ;)

**Comment by mohshbool** [comment]
> > please allow me to make those little suggestions 😃
> > i will continue the review process another time
> 
> @mohshbool all your suggestions committed! ;)

that's great! I have not finished reviewing, I'm sorry but I've been a bit busy, I will finish the review by midnight (utc+3) hopefully!

**Pull request merged**